### PR TITLE
SDK-316: Enable deploying O3 from build-distro plugin

### PR DESCRIPTION
**Pull request by wikumChamith**
jira ticket: https://issues.openmrs.org/browse/SDK-316

### CODE REVIEW

**Comment by dkayiwa** [comment]
@wikumChamith do we have wiki documentation on how this works?

**Comment by wikumChamith** [comment]
@dkayiwa we don't have an documentation on using this with O3.

**Comment by dkayiwa** [comment]
Do you think we should have it?

**Comment by wikumChamith** [comment]
> Do you think we should have it?

Do you think we should add it to the SDK documentation or the  O3 setup documentation?

**Comment by dkayiwa** [comment]
What do you think?

**Comment by wikumChamith** [comment]
> What do you think?

I think we should update the SDK documentation. We could also update both.

**Comment by dkayiwa** [comment]
> We could also update both.

Ok let us do it.

**Comment by wikumChamith** [comment]
@dkayiwa I updated the documentation: https://wiki.openmrs.org/pages/viewpage.action?pageId=224527013#O3ImplementerDocumentation:SetUp,Configure&Deploy-CreatingDockerConfigurationforOpenMRS3Distribution

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java [hunk 1]
```diff
@@ -64,6 +68,10 @@ public class BuildDistro extends AbstractTask {
 
 	private static final String DOCKER_COMPOSE_OVERRIDE_YML = "docker-compose.override.yml";
 
+	private static final String O2_Distribution = "2.x Distribution";
```
Comment (dkayiwa) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java:71]: https://wiki.openmrs.org/display/docs/Java+Conventions#JavaConventions-NamingConventions

Comment (wikumChamith) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java]: @dkayiwa I changed this according to the naming convention.

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java [hunk 2]
```diff
@@ -371,8 +407,7 @@ private void writeTemplatedFile(File targetDirectory, String distro, String vers
 		if (!compose.exists()) {
 			try (InputStream inputStream = composeUrl.openStream(); FileWriter composeWriter = new FileWriter(compose)) {
 				String content = IOUtils.toString(inputStream, StandardCharsets.UTF_8);
-				content = content.replaceAll("<distro>", distro);
-				content = content.replaceAll("<tag>", version);
```
Comment (dkayiwa) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java:375]: Can you throw in some light as to why you made this particular change?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java [hunk 3]
```diff
@@ -129,14 +137,43 @@ public void executeTask() throws MojoExecutionException, MojoFailureException {
...
+					if (DistroHelper.isRefapp2_3_1orLower(server.getDistroArtifactId(), server.getVersion())) {
+						distroProperties = new DistroProperties(server.getVersion());
+					} else {
+						distroProperties = distroHelper.downloadDistroProperties(buildDirectory, server);
+						distroArtifact = new Artifact(server.getDistroArtifactId(), server.getVersion(), server.getDistroGroupId(),
+								"jar");
+					}
+					break;
+				case O3_DISTRIBUTION:
+					wizard.promptForO3RefAppVersionIfMissing(server, versionsHelper);
```
Comment (dkayiwa) [inline maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java:157]: What you are doing for O3 looks more than what the `build-distro` command is supposed to be doing. For instance, why should i need docker to be running for this command to execute successfully?

File: maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java (hunk 2 above)
Comment (wikumChamith) [reply maven-plugin/src/main/java/org/openmrs/maven/plugins/BuildDistro.java:375]: In the new docker-compose file we used version is formated like {TAG:-nightly}.

**Comment by dkayiwa** [comment]
@wikumChamith did you see the above build failures?

**Comment by wikumChamith** [comment]
> @wikumChamith did you see the above build failures?

@dkayiwa I fixed the build errors.

**Comment by wikumChamith** [comment]
@dkayiwa  I tested this. This is working fine for me. Here are my logs.

build-distro: https://pastebin.com/EvxxrUFU
docker compose up: https://drive.google.com/file/d/1JMNIe9-DCOD1BWUYkJicgPB8nWTTr0XC/view?usp=drive_link

**Pull request merged**